draft-ietf-quic-transport.md

```diff
@@ -1224,41 +1225,77 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
-sequence number than the highest sequence number of any connection ID ever sent
-or received on that local address.  This ensures that when an endpoint migrates
-to a new path or changes connection ID on an existing path, the packets will use
-a new connection ID in both directions.
-
-Implementations SHOULD ensure that peers have a connection ID with a matching
-sequence number available when changing to a new connection ID.  An
-implementation could do this by always supplying a corresponding connection ID
-to a peer for each connection ID received from that peer.
-
-While endpoints select connection IDs as appropriate for their implementation,
-the connection ID MUST NOT include the unprotected sequence number.  Endpoints
-need to be able to recover the sequence number associated with each connection
-ID they generate without relying on information available to unaffiliated
-parties. A connection ID that encodes an unencrypted sequence number could be
-used to correlate connection IDs across network paths.
+An endpoint whose peer has selected a non-zero-length connection ID will
+potentially send packets using different connection IDs over the lifetime of a
+connection. These connection IDs are supplied by the peer using the
+NEW_CONNECTION_ID frame ({{frame-new-connection-id}}).
+
+
+### Issuing Connection IDs
+
+An endpoint issues connection IDs to its peer for the peer to use when sending
+packets. This allows the endpoint control over the strategy used to interpret
+connection IDs of packets that it receives.  These connection IDs are
+communicated to the peer using NEW_CONNECTION_ID frames
+({{frame-new-connection-id}}).
+
+The initial connection ID issued to the peer is sent by the endpoint as the
+Source Connection ID during the handshake, ensuring that the peer always begins
+the connection with at least one connection ID to use when sending.  An endpoint
+SHOULD supply its peer with additional connection IDs via NEW_CONNECTION_ID
+frames. While each endpoint can choose how many connection IDs to issue, a
+recommended number of outstanding connection IDs is eight.
+
+When an endpoint issues a connection ID, it MUST accept packets using this
+connection ID for the duration of the connection or until its peer invalidates
+the connection ID via a CONNECTION_ID_FINISHED frame
+({{frame-connection-id-finished}}).
+
+Implementations SHOULD ensure that peers have sufficient connection IDs
+available to reduce the possibility of peers exhausting their supply of
+available connection IDs.  An implementation could do this by always supplying a
+new connection ID for each connection ID retired with a CONNECTION_ID_FINISHED
+frame. When a receiver of a packet notices that its peer is now using a
+previously unused connection ID, it can choose to supply its peer with a new
+connection ID using a NEW_CONNECTION_ID frame to reduce the possibility of its
+peer running out of available connection IDs.
```

Comment:
Fair enough. I'm not sure what an endpoint would do with the knowledge that the peer implements this, it's more of an opportunistic thing in my mind: Either the endpoint has no CIDs available when it needs one or it has a CID it can use. 

If you're writing code that issues CIDs, you might want to think about this -- it could help keep your peer from running out when they need it. There are a couple of other recommendations of this sort that are designed to help things go smoothly, but aren't strictly necessary if you think you know better when you're implementing things. 

Either way, I definitely agree that we need to suggest peers issue new IDs as it gets _FINISHED frames (and I think we do today, unless I'm mistaken).

Reply:
> Either way, I definitely agree that we need to suggest peers issue new IDs as it gets _FINISHED frames (and I think we do today, unless I'm mistaken).

I agree that we suggest that, and my point is that we should only have that requirement.

The reason I prefer disallowing the peer to send additional CIDs before receiving _FINISHED frames is that such behavior requires the client to remember `max_probe_paths + 8` connection IDs. If we are to require that amount of memory to be reserved, why not raise the limit (i.e. 8)?

Reply:
Okay, thinking it through, I think I'm in agreement now. 
We don't ever require that a certain amount of memory is reserved on the person getting the NEW_ frames, they can happily ignore them and just work on a subset. The only potential problem there is if the issuer decides to retroactively lower the amount of CIDs that they maintain. 

So, if we remove this text and make the replacement on receipt of _FINISHED mandatory, then I think we're all set. 

And that keeps us on the same track with the "if you ever issue this many CIDs, you better be prepared to have that many CIDs you're dealing with forever" approach.

Reply:
I don't know if I agree though.  I would prefer to frame this differently:

1. you issue connection IDs as many as you feel like
2. you can't forget an issued connection ID unless your peer tells you they have forgotten it
3. you remember as many of the connection IDs you were given as you feel like and forget the rest
4. you can't reuse a connection ID for a different path, so forgetting it is a good idea
5. connection IDs you forget can be signaled with indicated with a FORGET_CONNECTION_ID frame if you need more or think you might need more

And then some advice:

* ideally a peer has as many connection IDs available to them as paths they might use, which is hard to predict, so allow for some spares (without multipath, 8 is probably plenty)
* as a peer tells you that is has forgotten connection IDs, it's best to issue them replacements so that they always have enough

Reply:
Generally, I'm completely on board with that. 

The only problem that would be solved by making sure that the other side *always* replaces when they get _FINISHED frames is this: 

1. Peer issues 16 CIDs
2. I only wanted 8, so I just ignore and throw out the rest
- Now normally this should work fine, we'd just all keep on going with 8 -
3. *If* the peer decides that they only want to maintain 8 and they just gave me 8 "extra" to start with, then I'm screwed if
4. I retire all 8 CIDs (or 7, or whatever)
5. The peer never sends me new ones

Now I'm stuck because they essentially retracted their giving of 16. 

We can solve all of that if we require replacement when _FINISHED.

Other alternative: Add a flag to _FINISHED saying "do/do not replace" and in step (2) above, I *tell* the peer that I'm throwing away 8. 
Not a huge fan of that, though. 

Or, and I'd be okay with this, we just say "the recommendation is there for a reason" and leave as is in the current text. Technically a peer could do that wrong, but they could also just not issue a lot of CIDs too, since 8 is only a recommendation, so we don't need to try to plug this hole and not the others.

Reply:
I think that this risk is easier acknowledged, but not addressed.  Yes, if you throw away half of the connection IDs, there's a chance that you will need them.  You took a risk and you might need to make a new connection to repair it.  That's not terrible.

Reply:
Yeah I like that perspective, let's leave this text for now

Reply:
I think I might agree to what @martinthomson says, assuming that we will not introduce a transport parameter that tells the peer how many CIDs the endpoint can remember.